pkg/matrix/matrix_types.go

```diff
@@ -0,0 +1,82 @@
+/*
+   Copyright 2022 The Tekton Authors
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package matrix
+
+import (
+	"strconv"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+// Combinations is a slice of combinations of Parameters from a Matrix.
+type Combinations []*Combination
+
+// Combination is a specific combination of Parameters from a Matrix.
+type Combination struct {
+	// MatrixID is an identification of a combination from Parameters in a Matrix.
+	MatrixID string
+
+	// Params is a specific combination of Parameters in a Matrix.
+	Params []v1beta1.Param
+}
+
+func (combinations Combinations) fanOut(param v1beta1.Param) Combinations {
+	if len(combinations) == 0 {
+		return initializeCombinations(param)
+	}
+	return combinations.distribute(param)
+}
+
+func (combinations Combinations) distribute(param v1beta1.Param) Combinations {
+	// when there are existing combinations, this is a non-first parameter in the matrix, and we need to distribute
+	// it among the existing combinations
+	var expandedCombinations Combinations
+	var count int
+	for _, value := range param.Value.ArrayVal {
+		for _, combination := range combinations {
+			expandedCombinations = append(expandedCombinations, createCombination(count, param.Name, value, combination.Params))
+			count++
+		}
+	}
+	return expandedCombinations
+}
+
+func initializeCombinations(param v1beta1.Param) Combinations {
+	// when there are no existing combinations, this is the first parameter in the matrix, so we initialize the
+	// combinations with the first Parameter
+	var combinations Combinations
+	for i, value := range param.Value.ArrayVal {
+		combinations = append(combinations, createCombination(i, param.Name, value, []v1beta1.Param{}))
+	}
+	return combinations
+}
+
+func createCombination(i int, name string, value string, parameters []v1beta1.Param) *Combination {
+	return &Combination{
+		MatrixID: strconv.Itoa(i),
+		Params: append(parameters, v1beta1.Param{
+			Name:  name,
+			Value: v1beta1.ArrayOrString{Type: v1beta1.ParamTypeString, StringVal: value},
+		}),
+	}
+}
+
+// ToMap converts a list of Combinations to a map where the key is the matrixId and the values are Parameters.
+func (combinations Combinations) ToMap() map[string][]v1beta1.Param {
```

Comment:
Can we define a `Combinations` as `map[string][]v1beta1.Param` instead of later converting it to map? I am trying to understand this conversion and how it will be used. 

_Please note that not asking to change in this PR - just wondering if we can simplify_

Reply:
@pritidesai - We had discussions about what information to surface in the pipelinerun about the combination used for a given taskrun created from a matrix. We [landed](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md#combinations-of-parameters-in-the-matrix) on providing the matrix only initially in line with providing the "minimal  information needed" as proposed in TEP-0100. Then, if we find that users need the mapping of matrixid to parameters, we can provide the complete information. The Combinations type with MatrixID and Parameters sets us up for this possible extension to align with the Kubernetes API convention of [subobjects over maps](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps). If we find that users don't need the complete information, we can switch to maps only given that it's only an implementation detail now.